File: Android/app/src/androidTest/java/jp/co/riso/smartdeviceapp/view/MainActivityTest.java
package jp.co.riso.smartdeviceapp.view;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import android.os.Bundle;
import android.os.Message;
import android.os.SystemClock;
import android.view.Gravity;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.test.platform.app.InstrumentationRegistry;

import org.junit.Test;

import jp.co.riso.smartdeviceapp.view.fragment.HomeFragment;
import jp.co.riso.smartprint.R;

public class MainActivityTest extends BaseActivityTestUtil {
    TestFragment mTestRightFragment = null;
    TestFragment mTestMainFragment = null;

    // sleep is needed because drawer calls goes through mHandler
    private void waitForDrawer() {
        SystemClock.sleep(1000);
    }

    // fragments for testing calls on right and main layout
    private void initLayoutFragments() {
        final FragmentManager fm = mainActivity.getSupportFragmentManager();

        mainActivity.runOnUiThread(() -> {
            fm.beginTransaction().add(R.id.rightLayout, new TestFragment()).commit();
            fm.executePendingTransactions();
        });
        InstrumentationRegistry.getInstrumentation().waitForIdleSync();

        Fragment rightFragment = fm.findFragmentById(R.id.rightLayout);
        assertTrue(rightFragment instanceof TestFragment);
        mTestRightFragment = (TestFragment) rightFragment;

        mainActivity.runOnUiThread(() -> {
            fm.beginTransaction().add(R.id.mainLayout, new TestFragment()).commit();
            fm.executePendingTransactions();
        });
        InstrumentationRegistry.getInstrumentation().waitForIdleSync();

        Fragment mainFragment = fm.findFragmentById(R.id.mainLayout);
        assertTrue(mainFragment instanceof TestFragment);
        mTestMainFragment = (TestFragment) mainFragment;
    }

    @Test
    public void testOpenClose_LeftDrawer() {
        mainActivity.openDrawer(Gravity.LEFT);
        waitForDrawer();
        assertTrue(mainActivity.isDrawerOpen(Gravity.LEFT));

        mainActivity.closeDrawers();
        waitForDrawer();
        assertFalse(mainActivity.isDrawerOpen(Gravity.LEFT));
    }

    @Test
    public void testOpenClose_RightDrawer() {
        initLayoutFragments();
        mainActivity.openDrawer(Gravity.RIGHT);
        waitForDrawer();
        assertTrue(mainActivity.isDrawerOpen(Gravity.RIGHT));
        assertTrue(mTestMainFragment.onPauseCalled);
        assertTrue(mTestRightFragment.onResumeCalled);

        mainActivity.closeDrawers();
        waitForDrawer();
        assertFalse(mainActivity.isDrawerOpen(Gravity.RIGHT));
        assertTrue(mTestMainFragment.onResumeCalled);
        assertTrue(mTestRightFragment.onPauseCalled);
    }

    @Test
    public void testOnBackPressed() {
        mainActivity.openDrawer(Gravity.LEFT);
        waitForDrawer();
        assertTrue(mainActivity.isDrawerOpen(Gravity.LEFT));

        mainActivity.onBackPressed();
        waitForDrawer();
        assertFalse(mainActivity.isDrawerOpen(Gravity.LEFT));

        initLayoutFragments();
        mainActivity.openDrawer(Gravity.RIGHT);
        waitForDrawer();
        assertTrue(mainActivity.isDrawerOpen(Gravity.RIGHT));

        mainActivity.onBackPressed();
        waitForDrawer();
        assertFalse(mainActivity.isDrawerOpen(Gravity.RIGHT));
    }

    @Test
    public void testOnSaveInstanceState_LeftDrawerOpen() {
        mainActivity.openDrawer(Gravity.LEFT);
        waitForDrawer();

        Bundle testBundle = new Bundle();
        mainActivity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                mainActivity.onSaveInstanceState(testBundle);
                assertTrue(testBundle.getBoolean(MainActivity.KEY_LEFT_OPEN));
                assertFalse(testBundle.getBoolean(MainActivity.KEY_RIGHT_OPEN));
                assertFalse(testBundle.getBoolean(MainActivity.KEY_RESIZE_VIEW));
            }
        });
    }

    @Test
    public void testOnSaveInstanceState_RightDrawerOpen() {
        initLayoutFragments();
        mainActivity.openDrawer(Gravity.RIGHT);
        waitForDrawer();

        Bundle testBundle = new Bundle();
        mainActivity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                mainActivity.onSaveInstanceState(testBundle);
                assertTrue(testBundle.getBoolean(MainActivity.KEY_RIGHT_OPEN));
                assertFalse(testBundle.getBoolean(MainActivity.KEY_LEFT_OPEN));
                assertFalse(testBundle.getBoolean(MainActivity.KEY_RESIZE_VIEW));
            }
        });
    }

    @Test
    public void testOnSaveInstanceState_ResizeView() {
        initLayoutFragments();
        mainActivity.openDrawer(Gravity.RIGHT, true);
        waitForDrawer();

        Bundle testBundle = new Bundle();
        mainActivity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                mainActivity.onSaveInstanceState(testBundle);
                assertFalse(testBundle.getBoolean(MainActivity.KEY_LEFT_OPEN));
                assertTrue(testBundle.getBoolean(MainActivity.KEY_RIGHT_OPEN));
                assertTrue(testBundle.getBoolean(MainActivity.KEY_RESIZE_VIEW));
            }
        });
    }

    @Test
    public void testStoreMessage() {
        Message msg = new Message();
        msg.what = 0; //MSG_OPEN_DRAWER
        assertTrue(mainActivity.storeMessage(msg));
        msg.what = 1; //MSG_CLOSE_DRAWER
        assertTrue(mainActivity.storeMessage(msg));
        msg.what = 2; //MSG_CLEAR_ICON_STATES
        assertTrue(mainActivity.storeMessage(msg));
        msg.what = -1; //others
        assertFalse(mainActivity.storeMessage(msg));
        msg.what = 3; //others
        assertFalse(mainActivity.storeMessage(msg));
    }

     public static class TestFragment extends HomeFragment {
        public boolean onResumeCalled = false;
        public boolean onPauseCalled = false;
        @Override
        public void onResume() {
            super.onResume();
            onResumeCalled = true;
        }
        @Override
        public void onPause() {
            super.onPause();
            onPauseCalled = true;
        }
    }
}

File: Android/app/src/androidTest/kotlin/jp/co/riso/smartdeviceapp/view/MainActivityTest.kt
package jp.co.riso.smartdeviceapp.view

import jp.co.riso.smartprint.R
import android.view.Gravity
import android.os.Bundle
import android.os.Message
import android.os.SystemClock
import androidx.test.platform.app.InstrumentationRegistry
import jp.co.riso.smartdeviceapp.view.fragment.HomeFragment
import org.junit.Assert
import org.junit.Test

class MainActivityTest : BaseActivityTestUtil() {
    private var _testRightFragment: TestFragment? = null
    private var _testMainFragment: TestFragment? = null

    // sleep is needed because drawer calls goes through mHandler
    private fun waitForDrawer() {
        SystemClock.sleep(1000)
    }

    // fragments for testing calls on right and main layout
    private fun initLayoutFragments() {
        val fm = mainActivity!!.supportFragmentManager
        mainActivity!!.runOnUiThread {
            fm.beginTransaction().add(R.id.rightLayout, TestFragment()).commit()
            fm.executePendingTransactions()
        }
        InstrumentationRegistry.getInstrumentation().waitForIdleSync()
        val rightFragment = fm.findFragmentById(R.id.rightLayout)
        Assert.assertTrue(rightFragment is TestFragment)
        _testRightFragment = rightFragment as TestFragment?
        mainActivity!!.runOnUiThread {
            fm.beginTransaction().add(R.id.mainLayout, TestFragment()).commit()
            fm.executePendingTransactions()
        }
        InstrumentationRegistry.getInstrumentation().waitForIdleSync()
        val mainFragment = fm.findFragmentById(R.id.mainLayout)
        Assert.assertTrue(mainFragment is TestFragment)
        _testMainFragment = mainFragment as TestFragment?
    }

    @Test
    fun testOpenClose_LeftDrawer() {
        mainActivity!!.openDrawer(Gravity.LEFT)
        waitForDrawer()
        Assert.assertTrue(mainActivity!!.isDrawerOpen(Gravity.LEFT))
        mainActivity!!.closeDrawers()
        waitForDrawer()
        Assert.assertFalse(mainActivity!!.isDrawerOpen(Gravity.LEFT))
    }

    @Test
    fun testOpenClose_RightDrawer() {
        initLayoutFragments()
        mainActivity!!.openDrawer(Gravity.RIGHT)
        waitForDrawer()
        Assert.assertTrue(mainActivity!!.isDrawerOpen(Gravity.RIGHT))
        Assert.assertTrue(_testMainFragment!!.onPauseCalled)
        Assert.assertTrue(_testRightFragment!!.onResumeCalled)
        mainActivity!!.closeDrawers()
        waitForDrawer()
        Assert.assertFalse(mainActivity!!.isDrawerOpen(Gravity.RIGHT))
        Assert.assertTrue(_testMainFragment!!.onResumeCalled)
        Assert.assertTrue(_testRightFragment!!.onPauseCalled)
    }

    @Test
    fun testOnBackPressed() {
        mainActivity!!.openDrawer(Gravity.LEFT)
        waitForDrawer()
        Assert.assertTrue(mainActivity!!.isDrawerOpen(Gravity.LEFT))
        mainActivity!!.onBackPressed()
        waitForDrawer()
        Assert.assertFalse(mainActivity!!.isDrawerOpen(Gravity.LEFT))
        initLayoutFragments()
        mainActivity!!.openDrawer(Gravity.RIGHT)
        waitForDrawer()
        Assert.assertTrue(mainActivity!!.isDrawerOpen(Gravity.RIGHT))
        mainActivity!!.onBackPressed()
        waitForDrawer()
        Assert.assertFalse(mainActivity!!.isDrawerOpen(Gravity.RIGHT))
    }

    @Test
    fun testOnSaveInstanceState_LeftDrawerOpen() {
        mainActivity!!.openDrawer(Gravity.LEFT)
        waitForDrawer()
        val testBundle = Bundle()
        mainActivity!!.runOnUiThread {
            mainActivity!!.onSaveInstanceState(testBundle)
            Assert.assertTrue(testBundle.getBoolean(MainActivity.KEY_LEFT_OPEN))
            Assert.assertFalse(testBundle.getBoolean(MainActivity.KEY_RIGHT_OPEN))
            Assert.assertFalse(testBundle.getBoolean(MainActivity.KEY_RESIZE_VIEW))
        }
    }

    @Test
    fun testOnSaveInstanceState_RightDrawerOpen() {
        initLayoutFragments()
        mainActivity!!.openDrawer(Gravity.RIGHT)
        waitForDrawer()
        val testBundle = Bundle()
        mainActivity!!.runOnUiThread {
            mainActivity!!.onSaveInstanceState(testBundle)
            Assert.assertTrue(testBundle.getBoolean(MainActivity.KEY_RIGHT_OPEN))
            Assert.assertFalse(testBundle.getBoolean(MainActivity.KEY_LEFT_OPEN))
            Assert.assertFalse(testBundle.getBoolean(MainActivity.KEY_RESIZE_VIEW))
        }
    }

    @Test
    fun testOnSaveInstanceState_ResizeView() {
        initLayoutFragments()
        mainActivity!!.openDrawer(Gravity.RIGHT, true)
        waitForDrawer()
        val testBundle = Bundle()
        mainActivity!!.runOnUiThread {
            mainActivity!!.onSaveInstanceState(testBundle)
            Assert.assertFalse(testBundle.getBoolean(MainActivity.KEY_LEFT_OPEN))
            Assert.assertTrue(testBundle.getBoolean(MainActivity.KEY_RIGHT_OPEN))
            Assert.assertTrue(testBundle.getBoolean(MainActivity.KEY_RESIZE_VIEW))
        }
    }

    @Test
    fun testStoreMessage() {
        val msg = Message()
        msg.what = 0 //MSG_OPEN_DRAWER
        Assert.assertTrue(mainActivity!!.storeMessage(msg))
        msg.what = 1 //MSG_CLOSE_DRAWER
        Assert.assertTrue(mainActivity!!.storeMessage(msg))
        msg.what = 2 //MSG_CLEAR_ICON_STATES
        Assert.assertTrue(mainActivity!!.storeMessage(msg))
        msg.what = -1 //others
        Assert.assertFalse(mainActivity!!.storeMessage(msg))
        msg.what = 3 //others
        Assert.assertFalse(mainActivity!!.storeMessage(msg))
    }

    class TestFragment : HomeFragment() {
        var onResumeCalled = false
        var onPauseCalled = false
        override fun onResume() {
            super.onResume()
            onResumeCalled = true
        }

        override fun onPause() {
            super.onPause()
            onPauseCalled = true
        }
    }
}